lib/peer.js

```diff
@@ -181,14 +191,72 @@ class Peer extends EventEmitter {
       this.destroy(err)
     })
 
+    wire.once('pe1', () => {
+      this.onPe1()
+    })
+    wire.once('pe2', () => {
+      this.onPe2()
+    })
+    wire.once('pe3', () => {
+      this.onPe3()
+    })
+    wire.once('pe4', () => {
+      this.onPe4()
+    })
     wire.once('handshake', (infoHash, peerId) => {
       this.onHandshake(infoHash, peerId)
     })
     this.startHandshakeTimeout()
 
     this.setThrottlePipes()
 
-    if (this.swarm && !this.sentHandshake) this.handshake()
+    if (this.swarm) {
+      if (this.type === 'tcpOutgoing') {
+        if (secure && this.retries === 0 && !this.sentPe1) this.sendPe1()
+        else if (!this.sentHandshake) this.handshake()
+      } else if (this.type !== 'tcpIncoming' && !this.sentHandshake) this.handshake()
+    }
+  }
+
+  sendPe1 () {
+    this.wire.sendPe1()
+    this.sentPe1 = true
+  }
+
+  onPe1 () {
+    this.sendPe2()
+  }
+
+  sendPe2 () {
+    this.wire.sendPe2()
+    this.sentPe2 = true
+  }
+
+  onPe2 () {
+    this.sendPe3()
+  }
+
+  sendPe3 () {
+    this.wire.sendPe3(this.swarm.infoHash)
+    this.sentPe3 = true
+  }
+
+  onPe3 (infoHashHash) {
+    if (this.swarm) {
+      if (this.swarm.infoHashHash !== infoHashHash) {
+        this.destroy(new Error('unexpected crypto handshake info hash for this swarm'))
+      }
+      this.sendPe4()
+    }
+  }
+
+  sendPe4 () {
+    this.wire.sendPe4(this.swarm.infoHash)
+    this.sentPe4 = true
+  }
+
+  onPe4 () {
+    if (!this.sentHandshake) this.handshake()
```

Comment:
man, this is like asking to use generator functions, I understand all PE's are sent in order?

Reply:
no, it's send in order by two pears - webtorrent may send 2 and 4 or 1 3 and 5
it's events from bittorent dht

Reply:
ah okay, that explains it